halo/attest/keeper/keeper_internal_test.go

```diff
@@ -0,0 +1,317 @@
+								{ValidatorAddress: val2.Address, Signature: val2.Bytes()},
+								{ValidatorAddress: val3.Address, Signature: val3.Bytes()},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				atts: []*Attestation{
+					{Id: 1, BlockRoot: blockRoot, ChainId: 1, Hash: blockHashes[0].Bytes(), Height: 500, Status: int32(Status_Pending)},
+				},
+				sigs: []*Signature{
+					{Id: 1, AttId: 1, Signature: val1.Bytes(), ValidatorAddress: val1.Address},
+					{Id: 2, AttId: 1, Signature: val2.Bytes(), ValidatorAddress: val2.Address},
+					{Id: 3, AttId: 1, Signature: val3.Bytes(), ValidatorAddress: val3.Address},
+				},
+			},
+		},
+		{
+			name: "add_same_vote_msg_twice",
+			args: args{
+				msg: &types.MsgAddVotes{
+					Authority: "test-authority",
+					Votes: []*types.AggVote{
+						{
+							BlockHeader: &types.BlockHeader{
+								ChainId: 1,
+								Height:  500,
+								Hash:    blockHashes[0].Bytes(),
+							},
+							BlockRoot: blockRoot,
+							Signatures: []*types.SigTuple{
+								{ValidatorAddress: val1.Address, Signature: val1.Bytes()},
+								{ValidatorAddress: val2.Address, Signature: val2.Bytes()},
+							},
+						},
+					},
+				},
+			},
+			prerequisites: []func(t *testing.T, k *Keeper, ctx sdk.Context){
+				func(t *testing.T, k *Keeper, ctx sdk.Context) {
+					t.Helper()
+					// the same message as the one in the args
+					msg := &types.MsgAddVotes{
+						Authority: "test-authority",
+						Votes: []*types.AggVote{
+							{
+								BlockHeader: &types.BlockHeader{
+									ChainId: 1,
+									Height:  500,
+									Hash:    blockHashes[0].Bytes(),
+								},
+								BlockRoot: blockRoot,
+								Signatures: []*types.SigTuple{
+									{ValidatorAddress: val1.Address, Signature: val1.Bytes()},
+									{ValidatorAddress: val2.Address, Signature: val2.Bytes()},
+								},
+							},
+						},
+					}
+					err := k.Add(ctx, msg)
+					require.NoError(t, err)
+				},
+			},
+			want: want{
+				atts: []*Attestation{
+					{Id: 1, BlockRoot: blockRoot, ChainId: 1, Hash: blockHashes[0].Bytes(), Height: 500, Status: int32(Status_Pending)},
+				},
+				sigs: []*Signature{
+					{Id: 1, AttId: 1, Signature: val1.Bytes(), ValidatorAddress: val1.Address},
+					{Id: 2, AttId: 1, Signature: val2.Bytes(), ValidatorAddress: val2.Address},
+				},
+			},
+		},
+		{
+			name: "mismatching_block_root",
+			args: args{
+				msg: &types.MsgAddVotes{
+					Authority: "test-authority",
+					Votes: []*types.AggVote{
+						{
+							BlockHeader: &types.BlockHeader{
+								ChainId: 1,
+								Height:  500,
+								Hash:    blockHashes[0].Bytes(),
+							},
+							BlockRoot: []byte("different root"), // the block root is intentionally different to cause an error
+							Signatures: []*types.SigTuple{
+								{ValidatorAddress: val1.Address, Signature: val1.Bytes()},
+								{ValidatorAddress: val2.Address, Signature: val2.Bytes()},
+							},
+						},
+					},
+				},
+			},
```

Comment:
these type literals are very verbose, contains a lot of duplication and hard to ready. What about using some concise helper functions to do the heavy lifting but makes the code easier to read? 

e.g.
```
Signatures: defaultSigs(val1, val2)
```

Reply:
Can even write a custom functional option based "builder" that returns a default happy case "MsgAggVote" that you can tune based on `WithX` functional options. Then it is super clear what the differences are between each of the test cases. At this point is isn't super clear what the differences are.

Reply:
I like this idea.

Reply:
As discussed during the standup - I'll address this in a follow up PR. It is my priority.